modules/peridynamics/src/materials/ComputeStrainBaseNOSPD.C

```diff
@@ -163,49 +161,52 @@ ComputeStrainBaseNOSPD::computeConventionalQpDeformationGradient()
+  std::vector<dof_id_type> bonds = _pdmesh.getBonds(_current_elem->node_id(_qp));
 
-  unsigned int nb_index =
-      std::find(neighbors.begin(), neighbors.end(), end_nd->id()) - neighbors.begin();
-  std::vector<dof_id_type> dg_neighbors = _pdmesh.getDefGradNeighbors(cur_nd->id(), nb_index);
+  dof_id_type nb_index =
+      std::find(neighbors.begin(), neighbors.end(), _current_elem->node_id(1 - _qp)) -
+      neighbors.begin();
+  std::vector<dof_id_type> dg_neighbors =
+      _pdmesh.getBondDeformationGradientNeighbors(_current_elem->node_id(_qp), nb_index);
+  Real dg_sub_vol = _pdmesh.getHorizonSubsetVolume(_current_elem->node_id(_qp), nb_index);
+  Real dg_sub_vol_sum = _pdmesh.getHorizonSubsetVolumeSum(_current_elem->node_id(_qp));
 
   // calculate the shape tensor and prepare the deformation gradient tensor
-  Real vol_nb, weight, dgnodes_vsum = 0.0;
+  Real vol_nb, weight_nb;
   RealGradient origin_vec_nb, current_vec_nb;
 
   for (unsigned int nb = 0; nb < dg_neighbors.size(); ++nb)
     if (_bond_status_var->getElementalValue(_pdmesh.elemPtr(bonds[dg_neighbors[nb]])) > 0.5)
     {
-      Node * dgneighbor_nb = _pdmesh.nodePtr(neighbors[dg_neighbors[nb]]);
-      vol_nb = _pdmesh.getPDNodeVolume(neighbors[dg_neighbors[nb]]);
-      dgnodes_vsum += vol_nb;
-      origin_vec_nb = *dgneighbor_nb - *_pdmesh.nodePtr(cur_nd->id());
+      vol_nb = _pdmesh.getNodeVolume(neighbors[dg_neighbors[nb]]);
+      origin_vec_nb = *_pdmesh.nodePtr(neighbors[dg_neighbors[nb]]) -
+                      *_pdmesh.nodePtr(_current_elem->node_id(_qp));
 
       for (unsigned int k = 0; k < _dim; ++k)
-        current_vec_nb(k) = origin_vec_nb(k) + _disp_var[k]->getNodalValue(*dgneighbor_nb) -
-                            _disp_var[k]->getNodalValue(*cur_nd);
+        current_vec_nb(k) =
+            origin_vec_nb(k) +
+            _disp_var[k]->getNodalValue(*_pdmesh.nodePtr(neighbors[dg_neighbors[nb]])) -
+            _disp_var[k]->getNodalValue(*_current_elem->node_ptr(_qp));
 
-      weight = _horiz_rad[_qp] / origin_vec_nb.norm();
+      weight_nb = _horizon_radius[_qp] / origin_vec_nb.norm();
       for (unsigned int k = 0; k < _dim; ++k)
       {
         for (unsigned int l = 0; l < _dim; ++l)
         {
-          _shape2[_qp](k, l) += weight * origin_vec_nb(k) * origin_vec_nb(l) * vol_nb;
+          _shape2[_qp](k, l) += weight_nb * origin_vec_nb(k) * origin_vec_nb(l) * vol_nb;
           _deformation_gradient[_qp](k, l) +=
-              weight * current_vec_nb(k) * origin_vec_nb(l) * vol_nb;
+              weight_nb * current_vec_nb(k) * origin_vec_nb(l) * vol_nb;
         }
         // calculate derivatives of deformation_gradient w.r.t displacements of node i
-        _ddgraddu[_qp](0, k) += -weight * origin_vec_nb(k) * vol_nb;
-        _ddgraddv[_qp](1, k) += -weight * origin_vec_nb(k) * vol_nb;
+        _ddgraddu[_qp](0, k) += -weight_nb * origin_vec_nb(k) * vol_nb;
```

Comment:
Do `-=` and get rid of the negative on `weight_nb`.

Reply:
The negative sign is due to differentiation. I think the current implementation would be more clear to keep the negative sign to the expression on the right.

Reply:
I don't really think I agree, but I'm not going to hold this up over that!